### PR TITLE
Implement the $bucket aggregation pipeline stage.

### LINKED ISSUE #441: $bucket operator for the aggregation pipeline is not implemented
I got the following error while running a query using `$bucket`.

```
NotImplementedError: $bucket is not a valid operator for the aggregation pipeline. See http://docs.mongodb.org/manual/meta/aggregation-quick-reference/ for a complete list of valid operators.
```

I believe https://github.com/mongomock/mongomock/blob/develop/Missing_Features.rst needs to get updated.

version: 3.11.1

### PR DESCRIPTION
**Pull request by pcorpet**
Fixes #441.

### CODE REVIEW

File: mongomock/collection.py [hunk 1]
```diff
@@ -1810,64 +1866,55 @@ def _extend_collection(out_collection, field, expression):
...
+                    out_collection = sorted(in_collection, key=lambda kv: kv[0])
+                    grouped = itertools.groupby(out_collection, lambda kv: kv[0])
+
+                    out_collection = []
+                    for (unused_key, doc_id), group in grouped:
+                        group_list = [kv[1] for kv in group]
+                        doc_dict = {'_id': doc_id}
+
+                        for field, value in iteritems(output_fields):
+                            for operator, key in iteritems(value):
```
Comment (zozoens31) [inline mongomock/collection.py]: It's unclear to me why you cannot put this loop inside the _accumulate_group function

File: mongomock/collection.py [hunk 2]
```diff
@@ -1810,64 +1866,55 @@ def _extend_collection(out_collection, field, expression):
...
 
                         grouped_collection.append(doc_dict)
 
                     out_collection = grouped_collection
 
+                elif k == '$bucket':
+                    group_by = v['groupBy']
+                    boundaries = v['boundaries']
+                    output_fields = v.get('output', {'count': {'$sum': 1}})
+                    default_value = v.get('default', None)
```
Comment (zozoens31) [inline mongomock/collection.py]: please, add tests for no default value

File: mongomock/collection.py [hunk 3]
```diff
@@ -1810,64 +1866,55 @@ def _extend_collection(out_collection, field, expression):
...
+                                    operator, key, group_list, doc_dict.get(field, None))
 
                         grouped_collection.append(doc_dict)
 
                     out_collection = grouped_collection
 
+                elif k == '$bucket':
+                    group_by = v['groupBy']
+                    boundaries = v['boundaries']
+                    output_fields = v.get('output', {'count': {'$sum': 1}})
```
Comment (zozoens31) [inline mongomock/collection.py]: please, add a test without an output object

File: tests/test__collection_api.py [hunk 4]
```diff
@@ -2099,3 +2099,103 @@ def test_aggregate_group_sum(self):
             'countNoData': 3,
         }]
         self.assertEqual(expect, list(actual))
+
+    def test__aggregate_bucket(self):
+        collection = self.db.collection
+        collection.drop()
+        collection.insert_many([
+            {
```
Comment (zozoens31) [inline tests/test__collection_api.py:2107]: please, add an element with exact value 200

**Review by zozoens31 — CHANGES_REQUESTED** [review]

File: mongomock/collection.py (hunk 1 above)
Comment (pcorpet) [reply mongomock/collection.py]: Done.

File: mongomock/collection.py (hunk 2 above)
Comment (pcorpet) [reply mongomock/collection.py]: Done.

File: tests/test__collection_api.py (hunk 4 above)
Comment (pcorpet) [reply tests/test__collection_api.py:2107]: Done.

File: mongomock/collection.py (hunk 3 above)
Comment (pcorpet) [reply mongomock/collection.py]: Done.

**Pull request merged**